classes/api.php

```diff
@@ -0,0 +1,317 @@
+     * This is done by splitting the blob into multiple blocks, and then combining them using a BlockList on the Azure side
+     * before finally setting the final md5 by setting the blob properties.
+     *
+     * @param string $key blob key
+     * @param StreamInterface $contentstream the blob contents as a stream
+     * @param string $md5 binary md5 hash of file contents. You likely need to call hex2bin before passing in here.
+     * @param string $contenttype Content type to set for the file.
+     * @return PromiseInterface Promise that resolves when complete. Note the response is NOT available here,
+     * because this operation involves many separate requests.
+     */
+    public function put_blob_multipart_async(string $key, StreamInterface $contentstream, string $md5,
+        string $contenttype = self::DEFAULT_CONTENT_TYPE): PromiseInterface {
+        // We make multiple calls to the Azure API to do multipart uploads, so wrap the entire thing
+        // into a single promise.
+        $entirepromise = new Promise(function() use (&$entirepromise, $key, $contentstream, $md5, $contenttype) {
+            // Split into blocks.
+            $counter = 0;
+            $blockids = [];
+            $promises = [];
+
+            while (true) {
+                $content = $contentstream->read(self::MULTIPART_BLOCK_SIZE);
+
+                // Finished reading, nothing more to upload.
+                if (empty($content)) {
+                    break;
+                }
+
+                // Each block has its own md5 specific to itself.
+                $blockmd5 = base64_encode(hex2bin(md5($content)));
+
+                // The block ID must be the same length regardles of the counter value.
+                // So pad them with zeros.
+                $blockid = base64_encode(
+                    str_pad($counter++, 6, '0', STR_PAD_LEFT)
+                );
+
+                $request = new Request('PUT', $this->build_blob_block_url($key, $blockid), ['content-md5' => $blockmd5], $content);
+                $promises[] = $this->client->sendAsync($request)->then(null, $this->clean_exception_sas_if_needed());
+                $blockids[] = $blockid;
+            };
+
+            if (count($blockids) > self::MAX_NUMBER_BLOCKS) {
+                throw new coding_exception("Max number of blocks reached, block size too small ?");
+            }
+
+            // Will throw exception if any fail - if any fail we want to abort early.
+            Utils::unwrap($promises);
+
+            // Commit the blocks together into a single blob.
+            $body = $this->make_block_list_xml($blockids);
+            $bodymd5 = base64_encode(hex2bin(md5($body)));
+            $request = new Request('PUT', $this->build_blocklist_url($key),
+                ['Content-Type' => 'application/xml', 'content-md5' => $bodymd5], $body);
+            $this->client->sendAsync($request)->then(null, $this->clean_exception_sas_if_needed())->wait();
```

Comment:
We are only using sync operations in general. Does is make sense to use the sync API instead?

Reply:
We need the async api so we can use `->then` to catch the exception and handle it - I don't think this is possible without standard try/catch which is a bit messier